### PR TITLE
rgw: add a default redirect field for zones

### PR DESCRIPTION
**Pull request by yehudasa**
We'll send a redirect response instead of a 404 in a zone that has
this configured. This is useful when setting a new zone that needs to
catch up with older zone. If an object/bucket is not found then we
redirect client to the other zone.

Signed-off-by: Yehuda Sadeh yehuda@redhat.com

### CODE REVIEW

**Comment by idealguo** [comment]
@yehudasa , this feature is welcome in our real product environment, do you have plan to merge it recently? may be we can help to test if you need.

**Comment by liewegas** [comment]
@yehudasa ping

**Comment by oritwas** [comment]
@yehudasa , please rebase

**Comment by oritwas** [comment]
@yehudasa , can you open a tracker issue?

File: src/rgw/rgw_common.h [hunk 1]
```diff
@@ -1250,6 +1250,7 @@ struct req_state {
   string zonegroup_endpoint;
   string bucket_instance_id;
   int bucket_instance_shard_id;
+  string redirect_zone_endpoint;
```
Comment (amitkumar50) [inline src/rgw/rgw_common.h]: @yehudasa would be good to change title from [DNM] rgw to 
rgw: add a default redirect field for zones``

File: src/rgw/rgw_rest.cc [hunk 2]
```diff
@@ -680,6 +680,23 @@ void end_header(struct req_state* s, RGWOp* op, const char *content_type,
   rgw_flush_formatter_and_reset(s, s->formatter);
 }
 
+static void build_redirect_url(req_state *s, const string& redirect_base, string *redirect_url)
+{
+  string& dest_uri = *redirect_url;
+  
+  dest_uri = redirect_base;
+  /*
+   * reqest_uri is always start with slash, so we need to remove
```
Comment (amitkumar50) [inline src/rgw/rgw_rest.cc]: @yehudasa `request_uri` instead of `reqest_uri`

**Comment by liewegas** [comment]
@yehudasa ping

**Comment by yehudasa** [comment]
@liewegas ack

**Comment by yehudasa** [comment]
@liewegas @oritwas @idealguo rebased, updated, and added a test.

**Comment by tchaikov** [comment]
retest this please.

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2017-11-25_00:12:48-rgw-wip-kefu-testing-2017-11-25-0120-distro-basic-mira/

File: src/rgw/rgw_admin.cc [hunk 3]
```diff
@@ -249,6 +249,7 @@ void usage()
   cout << "   --source-zone             specify the source zone (for data sync)\n";
   cout << "   --default                 set entity (realm, zonegroup, zone) as default\n";
   cout << "   --read-only               set zone as read-only (when adding to zonegroup)\n";
+  cout << "   --redirect-zone           specify zone id to redirect when response is 404 (not found)\n";
```
Comment (tchaikov) [inline src/rgw/rgw_admin.cc]:
```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/cli/radosgw-admin/help.t: failed
--- /home/jenkins-build/build/workspace/ceph-pull-requests/src/test/cli/radosgw-admin/help.t
+++ /home/jenkins-build/build/workspace/ceph-pull-requests/src/test/cli/radosgw-admin/help.t.err
@@ -200,6 +200,7 @@
      --source-zone             specify the source zone (for data sync)
      --default                 set entity (realm, zonegroup, zone) as default
      --read-only               set zone as read-only (when adding to zonegroup)
+     --redirect-zone           specify zone id to redirect when response is 404 (not found)
      --placement-id            placement id for zonegroup placement commands
      --tags=<list>             list of tags for zonegroup placement add and modify commands
      --tags-add=<list>         list of tags to add for zonegroup placement modify command
# Ran 1 tests, 0 skipped, 1 failed.
```

@yehudasa could you update the cli test accordingly?

**Comment by yehudasa** [comment]
@tchaikov updated

**Pull request merged**